homeassistant/components/unifi/sensor.py

```diff
@@ -74,88 +59,94 @@ def options_updated():
+            if client_id in controller.entities[DOMAIN][sensor_class.TYPE]:
                 continue
+            sensors.append(sensor_class(controller.api.clients[client_id], controller))
 
-            sensors[item_id] = sensor_class(
-                controller.api.clients[client_id], controller
-            )
-            new_sensors.append(sensors[item_id])
+    if sensors:
+        async_add_entities(sensors)
 
-    if new_sensors:
-        async_add_entities(new_sensors)
 
+class UniFiBandwidthSensor(UniFiClient):
+    """UniFi bandwidth sensor base class."""
 
-@callback
-def remove_entities(controller, mac_addresses, sensors, entity_registry):
-    """Remove select sensor entities."""
-    for mac in mac_addresses:
-
-        for direction in ("rx", "tx"):
-            item_id = f"{direction}-{mac}"
-
-            if item_id not in sensors:
-                continue
-
-            entity = sensors.pop(item_id)
-            controller.hass.async_create_task(entity.async_remove())
-
-
-class UniFiRxBandwidthSensor(UniFiClient):
-    """Receiving bandwidth sensor."""
-
-    @property
-    def state(self):
-        """Return the state of the sensor."""
-        if self._is_wired:
-            return self.client.wired_rx_bytes / 1000000
-        return self.client.raw.get("rx_bytes", 0) / 1000000
+    TYPE = ""
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Return the name of the client."""
-        name = self.client.name or self.client.hostname
-        return f"{name} RX"
+        return f"{super().name} {self.TYPE.upper()}"
 
     @property
-    def unique_id(self):
+    def unique_id(self) -> str:
         """Return a unique identifier for this bandwidth sensor."""
-        return f"rx-{self.client.mac}"
+        return f"{self.TYPE}-{self.client.mac}"
 
     @property
-    def unit_of_measurement(self):
+    def unit_of_measurement(self) -> str:
         """Return the unit of measurement of this entity."""
         return DATA_MEGABYTES
 
+    async def async_added_to_hass(self) -> None:
+        """Client entity created."""
+        await super().async_added_to_hass()
+        self.controller.entities[DOMAIN][self.TYPE].add(self.client.mac)
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, self.controller.signal_options_update, self.options_updated
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, self.controller.signal_remove, self.remove_item
+            )
+        )
 
-class UniFiTxBandwidthSensor(UniFiRxBandwidthSensor):
-    """Transmitting bandwidth sensor."""
+    async def async_will_remove_from_hass(self) -> None:
+        """Disconnect client object when removed."""
+        await super().async_will_remove_from_hass()
+        self.controller.entities[DOMAIN][self.TYPE].remove(self.client.mac)
+
+    @callback
+    def options_updated(self) -> None:
```

Comment:
I think the callbacks should be coroutines instead. Then we can await `self.async_remove` and save one task.

Reply:
I didn't know that was possible